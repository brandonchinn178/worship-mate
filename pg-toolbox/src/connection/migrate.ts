import { RunnerOption as NodePgMigrateOptions } from 'node-pg-migrate'

export type MigrateOptions = Partial<NodePgMigrateOptions> & {
  /**
   * If true, load arguments from process.argv as well. Useful for migration
   * scripts. Supports the following commands:
   *
   *   - up
   *   - up {N}
   *   - down
   *   - down {N}
   *   - redo
   *   - redo {N}
   */
  loadFromArgs?: boolean
}

export type MigrateDirection = 'up' | 'down'

export type MigrateDirectionOptions = Array<{
  direction: MigrateDirection
  count: number
}>

const parseMigrateCount = (arg: string | undefined) => {
  if (arg === undefined) {
    return Infinity
  }

  const count = parseInt(arg, 10)
  if (isNaN(count)) {
    throw new Error(`Invalid number: ${count}`)
  }

  return count
}

const parseMigrateDirections = (command: string): MigrateDirection[] => {
  switch (command) {
    case 'up':
      return ['up']
    case 'down':
      return ['down']
    case 'redo':
      return ['down', 'up']
    default: {
      throw new Error(`Unknown migrate command: ${command}`)
    }
  }
}

/**
 * Should be called with process.argv.
 */
export const parseMigrateArgs = (argv: string[]): MigrateDirectionOptions => {
  // [node executable, script path, ...args]
  const args = argv.slice(2)

  if (args[0] === undefined) {
    throw new Error('No migrate command specified')
  }

  const directions = parseMigrateDirections(args[0])

  const count = parseMigrateCount(args[1])

  return directions.map((direction) => ({ direction, count }))
}
