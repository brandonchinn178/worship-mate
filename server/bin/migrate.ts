#!/usr/bin/env ts-node

import { initDatabase } from '~/db'

const runMigrations = async () => {
  const db = initDatabase()
  const options = {
    checkOrder: true,
    loadFromArgs: true,
  }

  try {
    await db.migrate(options)
    console.log('Migrations complete')
  } finally {
    db.close()
  }
}

runMigrations().catch(console.error)
