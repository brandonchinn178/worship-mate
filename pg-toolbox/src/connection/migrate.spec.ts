import * as fc from 'fast-check'

import { loadCLIMigrateArgs } from './migrate'

const mockConsoleLog = jest.spyOn(console, 'log')

const mockProcessExit = (jest.spyOn(process, 'exit') as unknown) as jest.Mock

beforeEach(() => {
  jest.resetAllMocks()
  mockConsoleLog.mockImplementation(() => {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
  })
  mockProcessExit.mockImplementation(() => {
    throw new Error()
  })
})

afterEach(jest.resetAllMocks)

describe('loadCLIMigrateArgs', () => {
  it('parses an action and a count', () => {
    fc.assert(
      fc.property(
        fc.string(),
        fc.string(),
        fc.constantFrom('up', 'down', 'redo'),
        fc.integer(),
        (argv0, argv1, action, count) => {
          expect(
            loadCLIMigrateArgs([argv0, argv1, action, count.toString()]),
          ).toStrictEqual({ action, count })
        },
      ),
    )
  })

  it('returns Infinity with no count specified', () => {
    fc.assert(
      fc.property(
        fc.string(),
        fc.string(),
        fc.constantFrom('up', 'down', 'redo'),
        (argv0, argv1, command) => {
          const { count } = loadCLIMigrateArgs([argv0, argv1, command])
          expect(count).toBe(Infinity)
        },
      ),
    )
  })

  it('fails without a command', () => {
    fc.assert(
      fc.property(fc.string(), fc.string(), (argv0, argv1) => {
        expect(() => loadCLIMigrateArgs([argv0, argv1])).toThrow()
      }),
    )
  })

  it('fails with an unrecognized command', () => {
    fc.assert(
      fc.property(
        fc.string(),
        fc.string(),
        fc.string().filter((s) => !['up', 'down', 'redo'].includes(s)),
        (argv0, argv1, command) => {
          expect(() => loadCLIMigrateArgs([argv0, argv1, command])).toThrow()
        },
      ),
    )
  })

  it('prints a help message', () => {
    expect(() => {
      loadCLIMigrateArgs(['node', 'scripts/migrate.js', '--help'])
    }).toThrow()

    expect(mockConsoleLog).toHaveBeenCalled()
    const [usage] = mockConsoleLog.mock.calls[0]
    expect(usage).toMatchSnapshot()
  })

  it('fails with a non-numeric count', () => {
    fc.assert(
      fc.property(
        fc.string(),
        fc.string(),
        fc.constantFrom('up', 'down', 'redo'),
        fc.string().filter((s) => isNaN(parseInt(s, 10))),
        (argv0, argv1, command, count) => {
          expect(() => {
            loadCLIMigrateArgs([argv0, argv1, command, count])
          }).toThrow()
        },
      ),
    )
  })
})
