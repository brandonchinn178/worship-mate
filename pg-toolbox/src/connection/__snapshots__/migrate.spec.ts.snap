// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadCLIMigrateArgs prints a help message 1`] = `
"
Usage: scripts/migrate.js [up|down|redo] [COUNT]

Migrate the database with the given arguments.

Parameters:
  up|down|redo      The direction to migrate, with redo doing down first,
                    then up.

  COUNT             The number of migrations to run
"
`;
